code/modules/projectiles/projectile.dm

```diff
@@ -674,268 +735,412 @@
-	Range()
 
-/obj/projectile/proc/process_homing() //may need speeding up in the future performance wise.
+		var/distance_to_move = min(distance_to_border, pixels_to_move)
+		// For homing we cap the maximum distance to move every loop
+		if (homing && distance_to_move > SSprojectiles.pixels_per_decisecond)
+			distance_to_move = SSprojectiles.pixels_per_decisecond
+
+		// Figure out if we move to the next turf and if so, what its positioning relatively to us is
+		var/x_shift = SIGN(movement_vector.pixel_x) * (!isnull(x_to_border) && distance_to_move >= x_to_border)
+		var/y_shift = SIGN(movement_vector.pixel_y) * (!isnull(y_to_border) && distance_to_move >= y_to_border)
+		var/turf/new_turf = (x_shift || y_shift) ? locate(x + x_shift, y + y_shift, z) : loc
+		// Calculate where in the turf we will be when we cross the edge.
+		// This is a projectile variable because its also used in hit VFX
+		entry_x = pixel_x + movement_vector.pixel_x * distance_to_move - x_shift * ICON_SIZE_X
+		entry_y = pixel_y + movement_vector.pixel_y * distance_to_move - y_shift * ICON_SIZE_Y
+		var/delete_distance
+
+		if (x_shift || y_shift)
+			// We've hit an invalid turf, end of a z level or smth went wrong
+			if (!istype(new_turf))
+				qdel(src)
+				return
+
+			// Move to the next tile
+			step_towards(src, new_turf)
+			SEND_SIGNAL(src, COMSIG_PROJECTILE_MOVE_PROCESS_STEP)
+			// We hit something and got deleted, stop the loop
+			if (QDELETED(src))
+				return
+			// If we've impacted something, we need to animate our movement until the actual hit
+			// Otherwise the projectile visually disappears slightly before the actual impact
+			if (deletion_queued)
+				// distance_to_move is how much we have to step to get to the next turf, hypotenuse is how much we need
+				// to move in the next turf to get from entry to impact position
+				delete_distance = distance_to_move + sqrt((impact_x - entry_x) ** 2 + (impact_y - entry_y) ** 2)
+
+		// We cannot move more than one turf worth of distance per loop, so this is a safe solution
+		pixels_moved_last_tile += distance_to_move
+		if (!deletion_queued && pixels_moved_last_tile >= ICON_SIZE_ALL)
+			reduce_range()
+			if (QDELETED(src))
+				return
+			// Similarly with range out deletion, need to calculate how many pixels we can actually move before deleting
+			if (deletion_queued)
+				delete_distance = distance_to_move - (pixels_moved_last_tile - ICON_SIZE_ALL)
+
+		if (deletion_queued)
+			// We moved to the next turf first, then impacted something
+			// This means that we need to offset our visual position back to the previous turf, then figure out
+			// how much we moved on the next turf (or we didn't move at all in which case we both shifts are 0 anyways)
+			if (loc == new_turf)
+				pixel_x -= x_shift * ICON_SIZE_X
+				pixel_y -= y_shift * ICON_SIZE_Y
+
+			// Similarly to normal animate code, but use lowered deletion distance instead.
```

Comment:
this comment indicates otherwise which is why I ask (also like, isn't this wrong distance to deletion is always higher?)

Reply:
see above, the comment is correct

Reply:
it's higher for actually hitting a target at LEAST because we're literally adding, right?